Blocks/Persistence/test.cc

```diff
@@ -338,6 +350,103 @@ TEST(PersistenceLayer, File) {
   }
 }
 
+TEST(PersistenceLayer, FileDirectives) {
+  using namespace persistence_test;
+
+  using IMPL = current::persistence::File<StorableString>;
+
+  const std::string persistence_file_name = current::FileSystem::JoinPath(FLAGS_persistence_test_tmpdir, "data");
+  const auto file_remover = current::FileSystem::ScopedRmFile(persistence_file_name);
+
+  {
+    // An empty file - no entries and head equals -1us.
+    IMPL impl(persistence_file_name);
+    EXPECT_EQ(0u, impl.Size());
+    EXPECT_EQ(-1, impl.CurrentHead().count());
+    const auto head_idxts = impl.HeadAndLastPublishedIndexAndTimestamp();
+    ASSERT_EQ(false, Exists(head_idxts.idxts));
+    EXPECT_EQ(-1, head_idxts.head.count());
+  }
+
+  {
+    current::time::ResetToZero();
+
+    // A file consisting only of directives.
+    current::FileSystem::WriteStringToFile(
+        "#head 0000000000000000001\n"
+        "#unknown_directive\t\tblah\n",
+        persistence_file_name.c_str());
+    // Skip unknown directives.
+    IMPL impl(persistence_file_name);
+    EXPECT_EQ(1, impl.CurrentHead().count());
+    // Append a new head directive, because after the last one there was another directive.
+    current::time::SetNow(std::chrono::microseconds(2));
+    impl.UpdateHead();
+    EXPECT_EQ(2, impl.CurrentHead().count());
+    const auto head_idxts = impl.HeadAndLastPublishedIndexAndTimestamp();
+    ASSERT_EQ(false, Exists(head_idxts.idxts));
+    EXPECT_EQ(2, head_idxts.head.count());
+  }
+  EXPECT_EQ(
+      "#head 0000000000000000001\n"
+      "#unknown_directive\t\tblah\n"
+      "#head\t00000000000000000002\n",
+      current::FileSystem::ReadFileAsString(persistence_file_name));
+
+  {
+    current::time::ResetToZero();
+
+    current::FileSystem::WriteStringToFile(
+        "{\"index\":0,\"us\":100}\t{\"s\":\"foo\"}\n"
+        "{\"index\":1,\"us\":200}\t{\"s\":\"bar\"}\n"
+        "#head\t00000000000000000300\n"
+        "#some_other_directive\n"
+        "#head  00000000000000000400\n"
+        "{\"index\":2,\"us\":500}\t{\"s\":\"meh\"}\n"
+        "#head\t \t00000000000000000600\n",
```

Comment:
Why the directive lines aren't two-JSON-per-line-compliant? Something along `{“directive":"head","data":"0000...."}\t{}\n`
...